Feature: Group Actions

dscottboggs opened the pull request:
This PR provides access to the group controls from the REST endpoints.

dscottboggs committed:
Make CLI utitlity functions available globally

dscottboggs committed:
Add user delete route

dscottboggs committed:
Move config helpers out of the global scope

dscottboggs committed:
Implement most groups routes

dscottboggs committed:
Improve error response handling in tests

dscottboggs committed:
Add query delete route

dscottboggs committed:
Add delete route permission route

dscottboggs committed:
Final touches on group actions.

j8r commented:
The first commit for updating mysql can be dropped now, I have pushed a commit updating dependencies today.

j8r commented on spec/actions/groups_spec.cr:
```diff
@@ -1,34 +1,212 @@
+  struct ChangedGroup
+    include JSON::Serializable
+    property from : Config::Group
+    property to : Config::Group
```
Can this be informative names?
I propose `source`, `destination` or `target`.

j8r commented on src/cli.cr:
```diff
@@ -211,12 +211,17 @@ module DppmRestApi::CLI
 
+  # Require ALL variables
+  private macro require_all_args
+    required {{@def.args.splat}}
```
This is not really the best away to do this.
If the variables aren't optional, positional CLI arguments (`arguments`) can be used.
This isn't on this PR scope, I can submit one after.

dscottboggs replied:
Relevant changes have been reverted.

j8r reviewed (commented):
That's a big good PR :o
Can you explain the second commit?

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,171 @@
+    if body = context.request.body
+      Hash(String, Array(String)).from_json body
+    else
+      {} of String => Array(String)
```
```suggestion
     Hash(String, Array(String)).new
```
To be more consistent with the above types.

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,171 @@
+    index = DppmRestApi.permissions_config
+      .groups
+      .index { |grp| grp.id == group_id }
+    raise NoSuchGroup.new context, group_id_s if index.nil?
```
`find` would avoid to have to do a second lookup.
The above logic can also be in a method, that can be reused the other `delete` endpoints too.

dscottboggs replied:
I attempted the refactor to do this last night, and it didn't work. The problem is that since all of our data-types are structs, changing a found group doesn't update that group in the config, only the local copy of it. We need to find the index, get the value at the index, then reinsert the modified value back in... or switch to classes.

Thoughts?

j8r replied:
I can't be a struct issue, nothing is passed around, and most importantly Array is always a reference.
Here the deal is because it's an array, that's why an index is needed.

dscottboggs replied:
When you do...
```crystal
group = DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id } || raise "nil guard"
group.change_something
```
...then...
```crystal
DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id }.should_not eq group
```
...whereas if group were a class, they would still be equivalent. Does that make sense? Or am I mistaken?

j8r replied:
`Group#permissions` is a Hash, so a reference.
Try to use this, this will return a class instead of a copied struct:
```cr
DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id }
      .try &.permissions
```

j8r replied:
`Group#permissions` is a Hash, so a reference.
Try to use this, this will return a class instead of a copied struct:
```cr
DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id }
      .try &.permissions
```

j8r replied:
And what bout changing the name? The ID of a group should be immutable, but the name not.
Or, we can make `Group` totally immutable (no property), and we have to pass through `DppmRestApi::Config` to create one, that will check if the ID and the name already exist.

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,171 @@
+      response.field "successfullyModifiedGroup" do
+        response.object do
+          response.field("from") { old_group.to_json response }
+          response.field("to") { group.to_json response }
```
An API response should better have declarative names.
I suggest `source` and `target`

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,171 @@
+    # and overwrite the current value.
+    DppmRestApi.permissions_config.groups[index] = group
+    build_json context.response do |response|
+      response.field "successfullyModifiedGroup" do
```
I don't think it's needed, this is strange to have this as an object key.
If the response is 2**, we already know that the operation is successful.

dscottboggs replied:
Hm. Fair enough.

dscottboggs commented:
>  Can you explain the second commit?

I believe that is leftover from something that no longer applies, I will drop it.

dscottboggs committed:
Remove unnecessary response data

j8r commented:
Right, the response data is not needed :+1:

j8r commented:
You can drop the first commit updating MySQL, already in master

j8r commented on shard.lock:
```diff
@@ -62,7 +62,7 @@ shards:
   mysql:
     github: crystal-lang/crystal-mysql
-    version: 0.8.0
+    version: 0.9.0
```
You can drop the first commit updating MySQL, already in master

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,150 @@
+    group_id_s = context.params.url["id"]
+    group_id = group_id_s.to_i? || raise BadRequest.new context,
+      %[Group ID "#{group_id_s}" was not an integer!]
+    group_idx = DppmRestApi.permissions_config.groups.index { |grp| grp.id == group_id }
```
`reject!` can be used, then raise if the array is still the same.
This will also ensure that no other remaining group still have the ID.

dscottboggs replied:
This route isn't deleting the group, it's deleting access to this path for this group. `#reject!` wouldn't work... I could see `#map` like
```crystal
group_id_already_found = false
DppmRestApi.permissions_config.groups.map do |group|
  next group if group.id != group_id
  raise InternalServerError.new context, "duplicate group with id " + group_Id_s if group_id_already_found
  group_id_already_found = true
  group.remove_access_to_path context.params.path["path"]
  group
end
DppmRestApi.permissions_config.sync_to_disk
```
... or using `#update` like `DppmRestApi.permissions_config.groups.update group_idx do |group|...`

j8r replied:
Right, sorry

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,150 @@
+      %[Group ID "#{group_id_s}" was not an integer] if group_id.nil?
+    to_be_deleted = DppmRestApi.permissions_config
+      .groups
+      .index { |group| group.id == group_id }
```
But here you can use `reject!`

dscottboggs committed:
Refactor group modification routes to avoid indexing

Using the #map! and #reject! mutation methods instead of indexing and replacing at the index also should hopefully avoid race conditions as well when processing multiple parallel requests

j8r commented on src/actions/groups.cr:
```diff
@@ -101,50 +105,72 @@ module DppmRestApi::Actions::Groups
-    DppmRestApi.permissions_config.groups.delete_at to_be_deleted
+      .reject! do |group|
+        if group.id == group_id
+          raise DuplicateGroup.new context, group_id_s if group_already_found
```
I don't think it's good.
What does it mean for the user if (s)he gets this error, (s)he has to execute the request again?
This check of duplicated ID should be done at the group creation, only.

j8r commented on src/actions/groups.cr:
````diff
@@ -1,34 +1,176 @@
+  #
+  # So, in summary the request...
+  # ```
+  # /123/param/%2Fsome%2Fpath?some-param=value&param2&param3&param4=value2&param4=value3
````
`%2F`? You mean `/`?

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,176 @@
+        context.params.url["path"]
+      ) || raise NotFound.new context, "\
+                              no permissions found at the specified path \
+                              (possibly already deleted?)"
```
The phrase can start with an uppercase letter

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,176 @@
+    path = URI.unescape context.params.url["path"]
+    access_level = Access.parse?(context.params.url["access_level"])
+    raise InvalidAccessLevel.new context if access_level.nil?
+    DppmRestApi.permissions_config.groups.map! do |group|
```
Why recreating an Array?
We just need to get the permissions Hash of a group, then modifying this reference.

j8r commented on src/actions/groups.cr:
```diff
@@ -1,34 +1,176 @@
+    group_id = group_id_s.to_i? || raise BadRequest.new context,
+      %[Group ID "#{group_id_s}" was not an integer!]
+    group_already_found = false
+    DppmRestApi.permissions_config.groups.map do |group|
```
Same here,  `map` not needed

j8r committed:
Use a common fake path for specs

j8r commented:
Thanks @dscottboggs !
